File: frontend/validate-env.js
#!/usr/bin/env node
/**
 * Environment Variables Validation Script
 *
 * Validates that all required environment variables are set and properly formatted
 * for the Todo Web Application frontend.
 *
 * Usage:
 *     node validate-env.js
 *
 * Exit codes:
 *     0 - All validations passed
 *     1 - One or more validations failed
 */

const fs = require('fs');
const path = require('path');

// Color codes for terminal output
const colors = {
  reset: '\x1b[0m',
  bright: '\x1b[1m',
  red: '\x1b[91m',
  green: '\x1b[92m',
  yellow: '\x1b[93m',
  blue: '\x1b[94m',
};

// Helper functions for colored output
const printHeader = (text) => {
  console.log(`\n${colors.bright}${colors.blue}${'='.repeat(70)}${colors.reset}`);
  console.log(`${colors.bright}${colors.blue}${text.padStart((70 + text.length) / 2).padEnd(70)}${colors.reset}`);
  console.log(`${colors.bright}${colors.blue}${'='.repeat(70)}${colors.reset}\n`);
};

const printSuccess = (text) => {
  console.log(`${colors.green}✓${colors.reset} ${text}`);
};

const printWarning = (text) => {
  console.log(`${colors.yellow}⚠${colors.reset} ${text}`);
};

const printError = (text) => {
  console.log(`${colors.red}✗${colors.reset} ${text}`);
};

const printInfo = (text) => {
  console.log(`${colors.blue}ℹ${colors.reset} ${text}`);
};

/**
 * Validate API URL format
 * @param {string} url - API URL to validate
 * @returns {[boolean, string]} - [isValid, message]
 */
function validateApiUrl(url) {
  if (!url) {
    return [false, 'NEXT_PUBLIC_API_URL is not set'];
  }

  try {
    const parsed = new URL(url);

    // Check scheme
    if (!['http:', 'https:'].includes(parsed.protocol)) {
      return [false, `Invalid protocol '${parsed.protocol}'. Expected 'http:' or 'https:'`];
    }

    // Production warning
    if (parsed.protocol === 'http:' &&
        !parsed.hostname.includes('localhost') &&
        parsed.hostname !== '127.0.0.1') {
      return [false, 'Production API URL should use https://, not http://'];
    }

    // Check for trailing slash
    if (url.endsWith('/')) {
      printWarning('  API URL should not end with a trailing slash');
    }

    return [true, `Valid API URL: ${url}`];
  } catch (e) {
    return [false, `Invalid URL format: ${e.message}`];
  }
}

/**
 * Validate JWT secret strength
 * @param {string} secret - JWT secret to validate
 * @returns {[boolean, string]} - [isValid, message]
 */
function validateJwtSecret(secret) {
  if (!secret) {
    return [false, 'BETTER_AUTH_SECRET is not set'];
  }

  // Check for default/insecure values
  const insecureValues = [
    'your-secret-key-change-in-production',
    'changeme',
    'secret',
    'password',
  ];

  if (insecureValues.includes(secret.toLowerCase())) {
    return [false, 'BETTER_AUTH_SECRET is using a default/insecure value'];
  }

  // Check length
  if (secret.length < 32) {
    return [false, `BETTER_AUTH_SECRET is too short (${secret.length} chars). Minimum: 32 characters`];
  }

  // Check entropy
  const uniqueChars = new Set(secret.split('')).size;
  if (uniqueChars < 16) {
    return [false, `BETTER_AUTH_SECRET has low entropy (only ${uniqueChars} unique characters)`];
  }

  // Check character variety
  const hasLetter = /[a-zA-Z]/.test(secret);
  const hasNumber = /[0-9]/.test(secret);
  const hasSpecial = /[^a-zA-Z0-9]/.test(secret);

  if (!hasLetter || !hasNumber) {
    return [false, 'BETTER_AUTH_SECRET should contain letters, numbers, and optionally special characters'];
  }

  return [true, `Strong secret key (${secret.length} characters)`];
}

/**
 * Check all required environment variables
 * @returns {Object} - Validation results
 */
function checkEnvVars() {
  const results = {};

  printHeader('Environment Variables Validation');

  // Load .env.local if it exists
  const envFile = path.join(__dirname, '.env.local');
  if (fs.existsSync(envFile)) {
    printInfo(`Loading environment variables from: ${envFile}\n`);
    const envContent = fs.readFileSync(envFile, 'utf-8');
    envContent.split('\n').forEach(line => {
      line = line.trim();
      if (line && !line.startsWith('#')) {
        const [key, ...valueParts] = line.split('=');
        if (key) {
          process.env[key.trim()] = valueParts.join('=').trim();
        }
      }
    });
  } else {
    printWarning(`.env.local file not found at: ${envFile}`);
    printInfo('Checking system environment variables...\n');
  }

  // Required variables
  console.log(`${colors.bright}Required Variables:${colors.reset}\n`);

  const apiUrl = process.env.NEXT_PUBLIC_API_URL || '';
  const [apiUrlValid, apiUrlMsg] = validateApiUrl(apiUrl);
  if (apiUrlValid) {
    printSuccess(`NEXT_PUBLIC_API_URL: ${apiUrlMsg}`);
    results.NEXT_PUBLIC_API_URL = true;
  } else {
    printError(`NEXT_PUBLIC_API_URL: ${apiUrlMsg}`);
    results.NEXT_PUBLIC_API_URL = false;
  }

  const secret = process.env.BETTER_AUTH_SECRET || '';
  const [secretValid, secretMsg] = validateJwtSecret(secret);
  if (secretValid) {
    printSuccess(`BETTER_AUTH_SECRET: ${secretMsg}`);
    results.BETTER_AUTH_SECRET = true;
  } else {
    printError(`BETTER_AUTH_SECRET: ${secretMsg}`);
    results.BETTER_AUTH_SECRET = false;
  }

  return results;
}

/**
 * Print validation summary
 * @param {Object} results - Validation results
 */
function printSummary(results) {
  printHeader('Validation Summary');

  const total = Object.keys(results).length;
  const passed = Object.values(results).filter(v => v).length;
  const failed = total - passed;

  if (failed === 0) {
    printSuccess(`All ${total} validations passed!`);
    printInfo('\nYour environment is properly configured.');
  } else {
    printError(`${failed} of ${total} validations failed`);
    printInfo('\nPlease fix the errors above before running the application.');
  }
}

/**
 * Print security recommendations
 */
function printRecommendations() {
  printHeader('Security Recommendations');

  const recommendations = [
    '1. Never commit .env.local files to version control',
    '2. Use different secrets for development and production',
    '3. Generate JWT secrets with cryptographically secure random generators:',
    '   node -e "console.log(require(\'crypto\').randomBytes(32).toString(\'base64\'))"',
    '4. Ensure BETTER_AUTH_SECRET matches the backend configuration',
    '5. Use HTTPS for API URLs in production (never HTTP)',
    '6. Store secrets in secure environment variable systems (Vercel Secrets, etc.)',
    '7. Rotate secrets periodically (every 90 days)',
    '8. Never expose API keys or secrets in client-side code',
  ];

  recommendations.forEach(rec => printInfo(rec));
}

/**
 * Print example .env.local file
 */
function printExampleEnv() {
  printHeader('Example .env.local File');

  const example = `# API Configuration
NEXT_PUBLIC_API_URL=http://localhost:8000

# Authentication (must match backend)
BETTER_AUTH_SECRET=your-strong-secret-key-at-least-32-characters-long
`;

  console.log(example);
}

/**
 * Check if backend secret matches
 */
function checkBackendSecretMatch() {
  const frontendSecret = process.env.BETTER_AUTH_SECRET;
  if (!frontendSecret) return;

  const backendEnvFile = path.join(__dirname, '..', 'backend', '.env');
  if (!fs.existsSync(backendEnvFile)) {
    printWarning('\nCannot verify backend secret match: backend/.env not found');
    return;
  }

  const backendEnvContent = fs.readFileSync(backendEnvFile, 'utf-8');
  let backendSecret = '';

  backendEnvContent.split('\n').forEach(line => {
    line = line.trim();
    if (line.startsWith('BETTER_AUTH_SECRET=')) {
      backendSecret = line.split('=')[1].trim();
    }
  });

  if (backendSecret && frontendSecret !== backendSecret) {
    printHeader('Secret Mismatch Warning');
    printError('BETTER_AUTH_SECRET does not match between frontend and backend!');
    printInfo('Frontend and backend must use the same JWT secret for authentication to work.');
    printInfo('\nFrontend secret length: ' + frontendSecret.length);
    printInfo('Backend secret length: ' + backendSecret.length);
  } else if (backendSecret) {
    printHeader('Backend Verification');
    printSuccess('BETTER_AUTH_SECRET matches between frontend and backend');
  }
}

/**
 * Main validation function
 */
function main() {
  try {
    // Run validations
    const results = checkEnvVars();

    // Check backend secret match
    checkBackendSecretMatch();

    // Print summary
    printSummary(results);

    // Check if any validation failed
    const anyFailed = Object.values(results).some(v => !v);

    if (anyFailed) {
      printRecommendations();
      printExampleEnv();
      process.exit(1);
    }

    // Print recommendations even on success
    printRecommendations();

    process.exit(0);
  } catch (error) {
    printError(`Unexpected error: ${error.message}`);
    console.error(error.stack);
    process.exit(1);
  }
}

// Run if called directly
if (require.main === module) {
  main();
}

module.exports = {
  validateApiUrl,
  validateJwtSecret,
  checkEnvVars,
};

File: backend/validate_env.py
#!/usr/bin/env python3
"""
Environment Variables Validation Script

Validates that all required environment variables are set and properly formatted
for the Todo Web Application backend.

Usage:
    python validate_env.py

Exit codes:
    0 - All validations passed
    1 - One or more validations failed
"""

import os
import sys
import re
from typing import Dict, List, Tuple
from urllib.parse import urlparse


class Color:
    """Terminal color codes for output formatting."""
    GREEN = '\033[92m'
    YELLOW = '\033[93m'
    RED = '\033[91m'
    BLUE = '\033[94m'
    BOLD = '\033[1m'
    END = '\033[0m'


def print_header(text: str) -> None:
    """Print a formatted header."""
    print(f"\n{Color.BOLD}{Color.BLUE}{'=' * 70}{Color.END}")
    print(f"{Color.BOLD}{Color.BLUE}{text:^70}{Color.END}")
    print(f"{Color.BOLD}{Color.BLUE}{'=' * 70}{Color.END}\n")


def print_success(text: str) -> None:
    """Print a success message."""
    print(f"{Color.GREEN}✓{Color.END} {text}")


def print_warning(text: str) -> None:
    """Print a warning message."""
    print(f"{Color.YELLOW}⚠{Color.END} {text}")


def print_error(text: str) -> None:
    """Print an error message."""
    print(f"{Color.RED}✗{Color.END} {text}")


def print_info(text: str) -> None:
    """Print an info message."""
    print(f"{Color.BLUE}ℹ{Color.END} {text}")


def validate_database_url(url: str) -> Tuple[bool, str]:
    """
    Validate PostgreSQL database URL format.

    Args:
        url: Database connection URL

    Returns:
        Tuple of (is_valid, error_message)
    """
    if not url:
        return False, "DATABASE_URL is not set"

    try:
        parsed = urlparse(url)

        # Check scheme
        if parsed.scheme not in ['postgresql', 'postgresql+asyncpg']:
            return False, f"Invalid scheme '{parsed.scheme}'. Expected 'postgresql' or 'postgresql+asyncpg'"

        # Check hostname
        if not parsed.hostname:
            return False, "Missing hostname in DATABASE_URL"

        # Check database name
        if not parsed.path or parsed.path == '/':
            return False, "Missing database name in DATABASE_URL"

        # Check for SSL mode in production-like URLs
        if 'localhost' not in parsed.hostname and '127.0.0.1' not in parsed.hostname:
            if 'sslmode' not in url:
                return False, "Production database URL should include '?sslmode=require'"

        return True, "Valid PostgreSQL URL"

    except Exception as e:
        return False, f"Invalid URL format: {str(e)}"


def validate_jwt_secret(secret: str) -> Tuple[bool, str]:
    """
    Validate JWT secret key strength.

    Args:
        secret: JWT secret key

    Returns:
        Tuple of (is_valid, error_message)
    """
    if not secret:
        return False, "BETTER_AUTH_SECRET is not set"

    if secret in ['your-secret-key-change-in-production', 'changeme', 'secret']:
        return False, "BETTER_AUTH_SECRET is using a default/insecure value"

    if len(secret) < 32:
        return False, f"BETTER_AUTH_SECRET is too short ({len(secret)} chars). Minimum: 32 characters"

    # Check for sufficient entropy
    unique_chars = len(set(secret))
    if unique_chars < 16:
        return False, f"BETTER_AUTH_SECRET has low entropy (only {unique_chars} unique characters)"

    # Check if it's not just a simple word or phrase
    if secret.isalpha() or secret.isdigit():
        return False, "BETTER_AUTH_SECRET should contain a mix of characters (letters, numbers, symbols)"

    return True, f"Strong secret key ({len(secret)} characters)"


def validate_cors_origin(origin: str) -> Tuple[bool, str]:
    """
    Validate CORS origin format.

    Args:
        origin: CORS origin URL

    Returns:
        Tuple of (is_valid, error_message)
    """
    if not origin:
        return True, "Using default CORS origin (http://localhost:3000)"

    try:
        parsed = urlparse(origin)

        if not parsed.scheme:
            return False, "CORS_ORIGIN must include scheme (http:// or https://)"

        if parsed.scheme not in ['http', 'https']:
            return False, f"Invalid scheme '{parsed.scheme}'. Expected 'http' or 'https'"

        if not parsed.netloc:
            return False, "CORS_ORIGIN must include hostname"

        # Production warning
        if parsed.scheme == 'http' and 'localhost' not in parsed.netloc and '127.0.0.1' not in parsed.netloc:
            return False, "Production CORS_ORIGIN should use https://, not http://"

        return True, f"Valid CORS origin: {origin}"

    except Exception as e:
        return False, f"Invalid URL format: {str(e)}"


def validate_log_level(level: str) -> Tuple[bool, str]:
    """
    Validate logging level.

    Args:
        level: Log level string

    Returns:
        Tuple of (is_valid, error_message)
    """
    if not level:
        return True, "Using default log level (INFO)"

    valid_levels = ['DEBUG', 'INFO', 'WARNING', 'ERROR', 'CRITICAL']

    if level.upper() not in valid_levels:
        return False, f"Invalid log level '{level}'. Must be one of: {', '.join(valid_levels)}"

    if level.upper() == 'DEBUG':
        print_warning("  DEBUG log level should not be used in production")

    return True, f"Valid log level: {level.upper()}"


def check_required_env_vars() -> Dict[str, bool]:
    """
    Check all required environment variables.

    Returns:
        Dictionary mapping variable names to validation status
    """
    results = {}

    print_header("Environment Variables Validation")

    # Required variables
    required_vars = [
        ('DATABASE_URL', validate_database_url),
        ('BETTER_AUTH_SECRET', validate_jwt_secret),
    ]

    # Optional variables with validation
    optional_vars = [
        ('CORS_ORIGIN', validate_cors_origin),
        ('LOG_LEVEL', validate_log_level),
    ]

    print(f"{Color.BOLD}Required Variables:{Color.END}\n")

    for var_name, validator in required_vars:
        value = os.getenv(var_name, '')
        is_valid, message = validator(value)

        if is_valid:
            print_success(f"{var_name}: {message}")
            results[var_name] = True
        else:
            print_error(f"{var_name}: {message}")
            results[var_name] = False

    print(f"\n{Color.BOLD}Optional Variables:{Color.END}\n")

    for var_name, validator in optional_vars:
        value = os.getenv(var_name, '')
        is_valid, message = validator(value)

        if is_valid:
            print_success(f"{var_name}: {message}")
            results[var_name] = True
        else:
            print_warning(f"{var_name}: {message}")
            results[var_name] = False

    return results


def print_summary(results: Dict[str, bool]) -> None:
    """
    Print validation summary.

    Args:
        results: Dictionary of validation results
    """
    print_header("Validation Summary")

    total = len(results)
    passed = sum(1 for v in results.values() if v)
    failed = total - passed

    if failed == 0:
        print_success(f"All {total} validations passed!")
        print_info("\nYour environment is properly configured.")
    else:
        print_error(f"{failed} of {total} validations failed")
        print_info("\nPlease fix the errors above before running the application.")


def print_recommendations() -> None:
    """Print security recommendations."""
    print_header("Security Recommendations")

    recommendations = [
        "1. Never commit .env files to version control",
        "2. Use different secrets for development and production",
        "3. Generate JWT secrets with cryptographically secure random generators:",
        "   python -c \"import secrets; print(secrets.token_urlsafe(32))\"",
        "4. Rotate JWT secrets periodically (every 90 days)",
        "5. Use HTTPS in production (never HTTP)",
        "6. Enable database SSL/TLS (sslmode=require)",
        "7. Set LOG_LEVEL to INFO or WARNING in production (not DEBUG)",
        "8. Store secrets in secure secret management systems (AWS Secrets Manager, etc.)",
    ]

    for rec in recommendations:
        print_info(rec)


def print_example_env() -> None:
    """Print example .env file."""
    print_header("Example .env File")

    example = """# Database Configuration
DATABASE_URL=postgresql://user:password@localhost:5432/todo_db?sslmode=require

# Authentication
BETTER_AUTH_SECRET=your-strong-secret-key-at-least-32-characters-long

# CORS Configuration (optional)
CORS_ORIGIN=http://localhost:3000

# Logging Configuration (optional)
LOG_LEVEL=INFO
"""

    print(example)


def main() -> int:
    """
    Main validation function.

    Returns:
        Exit code (0 for success, 1 for failure)
    """
    try:
        # Load .env file if it exists
        env_file = os.path.join(os.path.dirname(__file__), '.env')
        if os.path.exists(env_file):
            print_info(f"Loading environment variables from: {env_file}\n")
            with open(env_file, 'r') as f:
                for line in f:
                    line = line.strip()
                    if line and not line.startswith('#'):
                        if '=' in line:
                            key, value = line.split('=', 1)
                            os.environ[key.strip()] = value.strip()
        else:
            print_warning(f".env file not found at: {env_file}")
            print_info("Checking system environment variables...\n")

        # Run validations
        results = check_required_env_vars()

        # Print summary
        print_summary(results)

        # Check if any required validation failed
        required_vars = ['DATABASE_URL', 'BETTER_AUTH_SECRET']
        required_failed = any(not results.get(var, False) for var in required_vars)

        if required_failed:
            print_recommendations()
            print_example_env()
            return 1

        # Print recommendations even on success
        print_recommendations()

        return 0

    except KeyboardInterrupt:
        print("\n\nValidation interrupted by user")
        return 1
    except Exception as e:
        print_error(f"Unexpected error: {str(e)}")
        return 1


if __name__ == "__main__":
    sys.exit(main())
